File: OnlyCue/Commands/CueNumberAutoFill.swift
import Foundation

/// Fills only the `nil` `cueNumber`s of a cue list so a grandMA2 push always has a
/// unique, ordered number per cue (#763). User-entered numbers are never changed:
/// each gap is filled integer-first (`[1, _, 3]` → `2`), falling back to a fractional
/// step from the lower bound when no whole number fits (`[1, _, 2]` → `1.1`). Cues are
/// processed in time order and each assigned value becomes the lower bound of the next,
/// so consecutive gaps stay strictly increasing and never collide with existing numbers.
enum CueNumberAutoFill {

    /// `cue.id -> assigned number` for the nil-numbered cues only. Numbered cues and an
    /// empty list produce no entries.
    static func assignments(for cues: [Cue]) -> [Cue.ID: Double] {
        // Stable time order (original index breaks time ties).
        let ordered = cues.enumerated()
            .sorted { lhs, rhs in
                lhs.element.time == rhs.element.time ? lhs.offset < rhs.offset : lhs.element.time < rhs.element.time
            }
            .map(\.element)

        // Seed the used set (integer-thousandths, matching the pre-flight's rounding).
        var used = Set(ordered.compactMap { $0.cueNumber.map(key) })

        // Upper bound for each slot = nearest *existing* number strictly after it.
        var nextExisting = [Double?](repeating: nil, count: ordered.count)
        var running: Double?
        for index in stride(from: ordered.count - 1, through: 0, by: -1) {
            nextExisting[index] = running
            if let number = ordered[index].cueNumber { running = number }
        }

        var result: [Cue.ID: Double] = [:]
        var lower: Double?  // nearest number before this slot (existing or just-assigned)
        for (index, cue) in ordered.enumerated() {
            guard cue.cueNumber == nil else { lower = cue.cueNumber; continue }
            // No representable unique number in this interval (e.g. a sub-thousandth gap or
            // numbers running backwards vs. time) → leave the cue nil for the pre-flight to
            // report, rather than fabricate a duplicate or an out-of-range value.
            guard let value = pick(lower: lower, upper: nextExisting[index], used: &used) else { continue }
            result[cue.id] = value
            lower = value
        }
        return result
    }

    /// Smallest legal MA2 cue number.
    private static let minNumber = 0.001

    private static func key(_ number: Double) -> Int { Int((number * 1000).rounded()) }

    private static func round3(_ number: Double) -> Double { (number * 1000).rounded() / 1000 }

    /// First free, legal number strictly between `lower` and `upper` (either may be open),
    /// preferring whole numbers, then a `+0.1` step from the lower bound, then a
    /// binary-subdivision fallback for a tight upper bound. Returns `nil` when the interval
    /// holds no unique thousandth `>= minNumber` — every candidate is validated by `fits`,
    /// so the result can never duplicate an existing number or fall out of range.
    private static func pick(lower: Double?, upper: Double?, used: inout Set<Int>) -> Double? {
        let base = lower ?? 0
        func fits(_ value: Double) -> Bool {
            value >= minNumber && value > base && (upper.map { value < $0 } ?? true) && !used.contains(key(value))
        }
        func take(_ value: Double) -> Double { used.insert(key(value)); return value }

        // Cue numbers are always positive, so floor(lower)+1 >= 1 (and 1 when open).
        var candidate = lower.map { Int(floor($0)) + 1 } ?? 1
        while upper.map({ Double(candidate) < $0 }) ?? true {
            if fits(Double(candidate)) { return take(Double(candidate)) }
            candidate += 1
        }

        // +0.1 steps from the lower bound (upper is non-nil here — the integer scan above
        // never exits for an open upper).
        var step = 1
        while let upper, round3(base + Double(step) * 0.1) < upper {
            let value = round3(base + Double(step) * 0.1)
            if fits(value) { return take(value) }
            step += 1
        }

        // Tight bound: subdivide (base, upper) looking for a free thousandth.
        guard let upper else { return nil }
        var high = upper
        for _ in 0..<40 {
            let mid = round3((base + high) / 2)
            if fits(mid) { return take(mid) }
            if high - base < minNumber { break }  // gap narrower than the thousandths grid
            high = mid
        }
        return nil
    }
}

File: windows/OnlyCue.Core/Planning/CueNumberAutoFill.cs
using OnlyCue.Core.Document;

namespace OnlyCue.Core.Planning;

/// <summary>
/// Fills only the <c>null</c> cue numbers of a cue list so a grandMA2 push always
/// has a unique, ordered number per cue. User-entered numbers are never changed:
/// each gap is filled integer-first (<c>[1, _, 3]</c> → <c>2</c>), falling back to
/// a fractional step from the lower bound when no whole number fits
/// (<c>[1, _, 2]</c> → <c>1.1</c>). Cues are processed in time order and each
/// assigned value becomes the lower bound of the next, so consecutive gaps stay
/// strictly increasing and never collide with existing numbers.
/// </summary>
/// <remarks>
/// A line-for-line re-implementation of the Swift <c>CueNumberAutoFill</c>
/// (<c>OnlyCue/Commands/CueNumberAutoFill.swift</c>), kept in lockstep by the
/// golden-vector contract (<c>golden/cue-numbering-v1.json</c>, epic #728 M1b).
///
/// Two details are load-bearing and easy to lose in translation:
/// <list type="bullet">
/// <item>Every rounding here is round-half-<b>away-from-zero</b>, matching Swift's
/// <c>.rounded()</c>. .NET's parameterless <c>Math.Round</c> is banker's rounding
/// and picks a different thousandth on exact midpoints.</item>
/// <item>The time sort tie-breaks on the cue's <i>original index</i>. Swift needs
/// that because <c>Array.sorted(by:)</c> is not a specified-stable sort; the same
/// explicit tie-break is reproduced here rather than leaning on
/// <c>OrderBy</c>'s stability, so the two cores agree by construction.</item>
/// </list>
/// </remarks>
public static class CueNumberAutoFill
{
    /// <summary>Smallest legal MA2 cue number.</summary>
    private const double MinNumber = 0.001;

    /// <summary>
    /// <c>cue.Id</c> → assigned number, for the null-numbered cues only. Numbered
    /// cues and an empty list produce no entries. A null-numbered cue can also be
    /// absent: when the interval around it holds no unique thousandth, it is left
    /// unassigned for the MA2 pre-flight to report rather than given a duplicate.
    /// </summary>
    public static IReadOnlyDictionary<Guid, double> Assignments(IReadOnlyList<Cue> cues)
    {
        // Stable time order (original index breaks time ties).
        var ordered = cues
            .Select((cue, index) => (Cue: cue, Index: index))
            .OrderBy(entry => entry, TimeThenIndex)
            .Select(entry => entry.Cue)
            .ToList();

        // Seed the used set (integer-thousandths, matching the pre-flight's rounding).
        var used = new HashSet<int>(
            ordered.Where(cue => cue.CueNumber.HasValue).Select(cue => Key(cue.CueNumber!.Value)));

        // Upper bound for each slot = nearest *existing* number strictly after it.
        var nextExisting = new double?[ordered.Count];
        double? running = null;
        for (var index = ordered.Count - 1; index >= 0; index--)
        {
            nextExisting[index] = running;
            if (ordered[index].CueNumber is { } number)
            {
                running = number;
            }
        }

        var result = new Dictionary<Guid, double>();
        double? lower = null;  // nearest number before this slot (existing or just-assigned)
        for (var index = 0; index < ordered.Count; index++)
        {
            var cue = ordered[index];
            if (cue.CueNumber is { } existing)
            {
                lower = existing;
                continue;
            }

            if (Pick(lower, nextExisting[index], used) is not { } value)
            {
                continue;
            }

            result[cue.Id] = value;
            lower = value;
        }

        return result;
    }

    private static readonly IComparer<(Cue Cue, int Index)> TimeThenIndex =
        Comparer<(Cue Cue, int Index)>.Create((lhs, rhs) =>
            lhs.Cue.Time == rhs.Cue.Time
                ? lhs.Index.CompareTo(rhs.Index)
                : lhs.Cue.Time.CompareTo(rhs.Cue.Time));

    private static int Key(double number) => (int)Math.Round(number * 1000, MidpointRounding.AwayFromZero);

    private static double Round3(double number) => Math.Round(number * 1000, MidpointRounding.AwayFromZero) / 1000;

    /// <summary>
    /// First free, legal number strictly between <paramref name="lower"/> and
    /// <paramref name="upper"/> (either may be open), preferring whole numbers,
    /// then a <c>+0.1</c> step from the lower bound, then a binary-subdivision
    /// fallback for a tight upper bound. Returns <c>null</c> when the interval
    /// holds no unique thousandth ≥ <see cref="MinNumber"/> — every candidate is
    /// validated by <c>Fits</c>, so the result can never duplicate an existing
    /// number or fall out of range.
    /// </summary>
    private static double? Pick(double? lower, double? upper, HashSet<int> used)
    {
        var baseValue = lower ?? 0;

        bool Fits(double value) =>
            value >= MinNumber
            && value > baseValue
            && (upper is not { } bound || value < bound)
            && !used.Contains(Key(value));

        double Take(double value)
        {
            used.Add(Key(value));
            return value;
        }

        // Cue numbers are always positive, so floor(lower)+1 >= 1 (and 1 when open).
        var candidate = lower is { } start ? (int)Math.Floor(start) + 1 : 1;
        while (upper is not { } ceiling || candidate < ceiling)
        {
            if (Fits(candidate))
            {
                return Take(candidate);
            }

            candidate += 1;
        }

        // +0.1 steps from the lower bound (upper is non-null here — the integer scan
        // above never exits for an open upper).
        var step = 1;
        while (upper is { } bound && Round3(baseValue + (step * 0.1)) < bound)
        {
            var value = Round3(baseValue + (step * 0.1));
            if (Fits(value))
            {
                return Take(value);
            }

            step += 1;
        }

        // Tight bound: subdivide (base, upper) looking for a free thousandth.
        if (upper is not { } high)
        {
            return null;
        }

        for (var attempt = 0; attempt < 40; attempt++)
        {
            var mid = Round3((baseValue + high) / 2);
            if (Fits(mid))
            {
                return Take(mid);
            }

            if (high - baseValue < MinNumber)
            {
                break;  // gap narrower than the thousandths grid
            }

            high = mid;
        }

        return null;
    }
}
